domain/user/state/state.go

```diff
@@ -0,0 +1,335 @@
+// does not exist or has been previously removed an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUserWithActivationKey(
+	ctx context.Context,
+	uuid domainuser.UUID,
+	user user.User,
+	creatorUUID domainuser.UUID,
+	activationKey string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		err = addUser(ctx, tx, uuid, user, creatorUUID)
+		if err != nil {
+			return err
+		}
+		return setActivationKey(ctx, tx, uuid, activationKey)
+	})
+	if err != nil {
+		return fmt.Errorf("cannot add user %q with activation key: %w", user.Name, err)
+	}
+
+	return nil
+}
+
+// GetUser will retrieve the user specified by name from the database where
+// the user is active and has not been removed. If the user does not exist
+// or is deleted an error that satisfies usererrors.NotFound will be
+// returned.
+func (st *State) GetUser(ctx context.Context, userUUID domainuser.UUID) (domainuser.User, error) {
+	db, err := st.DB()
+	if err != nil {
+		return domainuser.User{}, fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	var domainUser domainuser.User
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		domainUser, err = getUser(ctx, tx, userUUID)
+		return err
+	})
+
+	return domainUser, err
+}
+
+// RemoveUser marks the user as removed. This obviates the ability of a user
+// to function, but keeps the user retaining provenance, i.e. auditing.
+// RemoveUser will also remove any credentials and activation codes for the
+// user. If no user exists for the given name then an error that satisfies
+// usererrors.NotFound will be returned.
+func (st *State) RemoveUser(ctx context.Context, userUUID domainuser.UUID) error {
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	removeUserQuery := "UPDATE user SET removed = true WHERE uuid = ?"
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		_, err = tx.ExecContext(ctx, removeUserQuery, userUUID.String())
+		return err
+	})
+
+	return err
+}
+
+// SetActivationKey removes any active passwords for the user and sets the
+// activation key. If no user is found for the supplied name an error
+// is returned that satisfies usererrors.NotFound.
+func (st *State) SetActivationKey(ctx context.Context, userUUID domainuser.UUID, activationKey string) error {
+	removed, err := st.isRemoved(ctx, userUUID)
+	if err != nil {
+		return fmt.Errorf("cannot get user %q: %w", userUUID, err)
+	}
+	if removed {
+		return fmt.Errorf("cannot set activation key for removed user %q", userUUID)
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		err = removePasswordHash(ctx, tx, userUUID)
```

Comment:
Ah sorry I missed that you were doing it this way for my comments above about removing password hash. I'll update my comments now. Thanks and nice job :)

Reply:
good :)